login/service.go

```diff
@@ -0,0 +1,163 @@
+package login
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/almighty/almighty-core/account"
+	"github.com/almighty/almighty-core/app"
+	"github.com/almighty/almighty-core/token"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+// Service defines the basic entrypoint required to perform a remote oauth login
+type Service interface {
+	Perform(ctx *app.AuthorizeLoginContext) error
+}
+
+// NewGitHubOAuth creates a new login.Service capable of using GitHub for authorization
+func NewGitHubOAuth(config *oauth2.Config, identities account.IdentityRepository, users account.UserRepository, tokenManager token.Manager) Service {
+	return &gitHubOAuth{
+		config:       config,
+		identities:   identities,
+		users:        users,
+		tokenManager: tokenManager,
+	}
+}
+
+type gitHubOAuth struct {
+	config       *oauth2.Config
+	identities   account.IdentityRepository
+	users        account.UserRepository
+	tokenManager token.Manager
+}
+
+// TEMP: This will leak memory in the long run with many 'failed' redirect attemts
+var stateReferer = map[string]string{}
```

Comment:
When we update the design for this, we need to take care of a couple of things :

1. This has to be made thread-safe.
2. When ALM Core runs on multiple nodes, this `stateReferer` information needs to be stored in a cache which is common for all nodes, otherwise we have to implement sticky session at the proxy level for authorize URLs.